src/hotspot/share/gc/g1/g1EvacFailureObjectsSet.cpp

```diff
@@ -0,0 +1,131 @@
+#include "utilities/quickSort.hpp"
+
+
+const G1SegmentedArrayAllocOptions G1EvacFailureObjectsSet::_alloc_options =
+  G1SegmentedArrayAllocOptions((uint)sizeof(OffsetInRegion), BufferLength, UINT_MAX, Alignment);
+
+G1SegmentedArrayBufferList<mtGC> G1EvacFailureObjectsSet::_free_buffer_list;
+
+#ifdef ASSERT
+void G1EvacFailureObjectsSet::assert_is_valid_offset(size_t offset) const {
+  const uint max_offset = 1u << (HeapRegion::LogOfHRGrainBytes - LogHeapWordSize);
+  assert(offset < max_offset, "must be, but is " SIZE_FORMAT, offset);
+}
+#endif
+
+oop G1EvacFailureObjectsSet::from_offset(OffsetInRegion offset) const {
+  assert_is_valid_offset(offset);
+  return cast_to_oop(_bottom + offset);
+}
+
+G1EvacFailureObjectsSet::OffsetInRegion G1EvacFailureObjectsSet::to_offset(oop obj) const {
+  const HeapWord* o = cast_from_oop<const HeapWord*>(obj);
+  size_t offset = pointer_delta(o, _bottom);
+  assert(obj == from_offset(static_cast<OffsetInRegion>(offset)), "must be");
+  return static_cast<OffsetInRegion>(offset);
+}
+
+G1EvacFailureObjectsSet::G1EvacFailureObjectsSet(uint region_idx, HeapWord* bottom) :
+  DEBUG_ONLY(_region_idx(region_idx) COMMA)
+  _bottom(bottom),
+  _offsets(&_alloc_options, &_free_buffer_list)  {
+  assert(HeapRegion::LogOfHRGrainBytes < 32, "must be");
+}
+
+void G1EvacFailureObjectsSet::record(oop obj) {
+  assert(obj != NULL, "must be");
+  assert(_region_idx == G1CollectedHeap::heap()->heap_region_containing(obj)->hrm_index(), "must be");
+  OffsetInRegion* e = _offsets.allocate();
+  *e = to_offset(obj);
+}
+
+// Helper class to join, sort and iterate over the previously collected segmented
+// array of objects that failed evacuation.
+class G1EvacFailureObjectsIterationHelper {
+  typedef G1EvacFailureObjectsSet::OffsetInRegion OffsetInRegion;
+
+  G1EvacFailureObjectsSet* _objects_set;
+  const G1SegmentedArray<OffsetInRegion, mtGC>* _segments;
+  OffsetInRegion* _offset_array;
+  uint _array_length;
+
+  static int order_oop(OffsetInRegion a, OffsetInRegion b) {
+    return static_cast<int>(a-b);
+  }
+
+  void join_and_sort() {
+    uint num = _segments->num_allocated_nodes();
+    _offset_array = NEW_C_HEAP_ARRAY(OffsetInRegion, num, mtGC);
+
+    _segments->iterate_nodes(*this);
+    assert(_array_length == num, "must be %u, %u", _array_length, num);
+
+    QuickSort::sort(_offset_array, _array_length, order_oop, true);
+  }
+
+  void iterate_internal(ObjectClosure* closure) {
+    for (uint i = 0; i < _array_length; i++) {
+      oop cur = _objects_set->from_offset(_offset_array[i]);
+      closure->do_object(cur);
+    }
+
+    FREE_C_HEAP_ARRAY(OffsetInRegion, _offset_array);
+  }
+
+public:
+  G1EvacFailureObjectsIterationHelper(G1EvacFailureObjectsSet* collector) :
+    _objects_set(collector),
+    _segments(&_objects_set->_offsets),
+    _offset_array(nullptr),
+    _array_length(0) { }
+
+  void iterate(ObjectClosure* closure) {
+    join_and_sort();
+    iterate_internal(closure);
```

Comment:
I would probably move the array allocation and freeing here instead of having this at the start and end of `join_and_sort` and `iterate_internal` respectively. Otherwise there is a hidden dependency on the first method allocating and the second freeing it, and looks cleaner as allocation and deallocation is obvious and on the same call level.

I.e.

```
  void iterate(ObjectClosure* closure) {
     uint num = _segments->num_allocated_nodes();
    _offset_array = NEW_C_HEAP_ARRAY(OffsetInRegion, num, mtGC);

    join_and_sort();
    iterate_internal(closure);

    FREE_C_HEAP_ARRAY(OffsetInRegion, _offset_array);
  }
```

Reply:
Thanks Thomas, good catch.